[lldb] Add support for large watchpoints in lldb

jasonmolenda opened the pull request:
This patch is the next piece of work in my Large Watchpoint proposal, https://discourse.llvm.org/t/rfc-large-watchpoint-support-in-lldb/72116

This patch breaks a user's watchpoint into one or more WatchpointResources which reflect what the hardware registers can cover. This means we can watch objects larger than 8 bytes, and we can watched unaligned address ranges.  On a typical 64-bit target with 4 watchpoint registers you can watch 32 bytes of memory if the start address is doubleword aligned.

Additionally, if the remote stub implements AArch64 MASK style watchpoints (e.g. debugserver on Darwin), we can watch any power-of-2 size region of memory up to 2GB, aligned to that same size.

I updated the Watchpoint constructor and CommandObjectWatchpoint to create a CompilerType of Array<UInt8> when the size of the watched region is greater than pointer-size and we don't have a variable type to use.  For pointer-size and smaller, we can display the watched granule as an integer value; for larger-than-pointer-size we will display as an array of bytes.

I have `watchpoint list` now print the WatchpointResources used to implement the watchpoint.

I added a WatchpointAlgorithm class which has a top-level static method that takes an enum flag mask WatchpointHardwareFeature and a user address and size, and returns a vector of WatchpointResources covering the request.  It does not take into account the number of watchpoint registers the target has, or the number still available for use.  Right now there is only one algorithm, which monitors power-of-2 regions of memory.  For up to pointer-size, this is what Intel hardware supports. AArch64 Byte Address Select watchpoints can watch any number of contiguous bytes in a pointer-size memory granule, that is not currently supported so if you ask to watch bytes 3-5, the algorithm will watch the entire doubleword (8 bytes). The newly default "modify" style means we will silently ignore modifications to bytes outside the watched range.

I've temporarily skipped TestLargeWatchpoint.py for all targets. It was only run on Darwin when using the in-tree debugserver, which was a proxy for "debugserver supports MASK watchpoints".  I'll be adding the aforementioned feature flag from the stub and enabling full mask watchpoints when a debugserver with that feature is enabled, and re-enable this test.

I added a new TestUnalignedLargeWatchpoint.py which only has one test but it's a great one, watching a 22-byte range that is unaligned and requires four 8-byte watchpoints to cover.

I also added a unit test, WatchpointAlgorithmsTests, which has a number of simple tests against WatchpointAlgorithms::PowerOf2Watchpoints. I think there's interesting possible different approaches to how we cover these; I note in the unit test that a user requesting a watch on address 0x12e0 of 120 bytes will be covered by two watchpoints today, a 128-bytes at 0x1280 and at 0x1300.  But it could be done with a 16-byte watchpoint at 0x12e0 and a 128-byte at 0x1300, which would have fewer false positives/private stops.  As we try refining this one, it's helpful to have a collection of tests to make sure things don't regress.

I tested this on arm64 macOS, (genuine) x86_64 macOS, and AArch64 Ubuntu.  I have not modifed the Windows process plugins yet, I might try that as a standalone patch, I'd be making the change blind, but the necessary changes (see ProcessGDBRemote::EnableWatchpoint) are pretty small so it might be obvious enough that I can change it and see what the Windows CI thinks.

There isn't yet a packet (or a qSupported feature query) for the gdb remote serial protocol stub to communicate its watchpoint capabilities to lldb.  I'll be doing that in a patch right after this is landed, having debugserver advertise its capability of AArch64 MASK watchpoints, and have ProcessGDBRemote add eWatchpointHardwareArmMASK to WatchpointAlgorithms so we can watch larger than 32-byte requests on Darwin.

I haven't yet tackled WatchpointResource *sharing* by multiple Watchpoints.  This is all part of the goal, especially when we may be watching a larger memory range than the user requested, if they then add another watchpoint next to their first request, it may be covered by the same WatchpointResource (hardware watchpoint register). Also one "read" watchpoint and one "write" watchpoint on the same memory granule need to be handled, making the WatchpointResource cover all requests.

As WatchpointResources aren't shared among multiple Watchpoints yet, there's no handling of running the conditions/commands/etc on multiple Watchpoints when their shared WatchpointResource is hit. The goal beyond "large watchpoint" is to unify (much more) the Watchpoint and Breakpoint behavior and commands.  I have a feeling I may be slowly chipping away at this for a while.

rdar://108234227

jasonmolenda committed:
[lldb] Add support for large watchpoints in lldb

This patch is the next piece of work in my Large Watchpoint proposal,
https://discourse.llvm.org/t/rfc-large-watchpoint-support-in-lldb/72116

This patch breaks a user's watchpoint into one or more WatchpointResources
which reflect what the hardware registers can cover. This means we
can watch objects larger than 8 bytes, and we can watched unaligned
address ranges.  On a typical 64-bit target with 4 watchpoint
registers you can watch 32 bytes of memory if the start address is
doubleword aligned.

Additionally, if the remote stub implements AArch64 MASK style
watchpoints (e.g. debugserver on Darwin), we can watch any power-of-2
size region of memory up to 2GB, aligned to that same size.

I updated the Watchpoint constructor and CommandObjectWatchpoint
to create a CompilerType of Array<UInt8> when the size of the watched
region is greater than pointer-size and we don't have a variable
type to use.  For pointer-size and smaller, we can display the
watched granule as an integer value; for larger-than-pointer-size
we will display as an array of bytes.

I have `watchpoint list` now print the WatchpointResources used to
implement the watchpoint.

I added a WatchpointAlgorithm class which has a top-level static
method that takes an enum flag mask WatchpointHardwareFeature and
a user address and size, and returns a vector of WatchpointResources
covering the request.  It does not take into account the number of
watchpoint registers the target has, or the number still available
for use.  Right now there is only one algorithm, which monitors
power-of-2 regions of memory.  For up to pointer-size, this is what
Intel hardware supports. AArch64 Byte Address Select watchpoints
can watch any number of contiguous bytes in a pointer-size memory
granule, that is not currently supported so if you ask to watch
bytes 3-5, the algorithm will watch the entire doubleword (8 bytes).
The newly default "modify" style means we will silently ignore
modifications to bytes outside the watched range.

I've temporarily skipped TestLargeWatchpoint.py for all targets.
It was only run on Darwin when using the in-tree debugserver, which
was a proxy for "debugserver supports MASK watchpoints".  I'll be
adding the aforementioned feature flag from the stub and enabling
full mask watchpoints when a debugserver with that feature is enabled,
and re-enable this test.

I added a new TestUnalignedLargeWatchpoint.py which only has one
test but it's a great one, watching a 22-byte range that is unaligned
and requires four 8-byte watchpoints to cover.

I also added a unit test, WatchpointAlgorithmsTests, which has a
number of simple tests against WatchpointAlgorithms::PowerOf2Watchpoints.
I think there's interesting possible different approaches to how
we cover these; I note in the unit test that a user requesting a
watch on address 0x12e0 of 120 bytes will be covered by two watchpoints
today, a 128-bytes at 0x1280 and at 0x1300.  But it could be done
with a 16-byte watchpoint at 0x12e0 and a 128-byte at 0x1300, which
would have fewer false positives/private stops.  As we try refining
this one, it's helpful to have a collection of tests to make sure
things don't regress.

I tested this on arm64 macOS, x86_64 macOS, and AArch64 Ubuntu.  I
have not modifed the Windows process plugins yet, I might try that
as a standalone patch, I'd be making the change blind, but the
necessary changes (see ProcessGDBRemote::EnableWatchpoint) are
pretty small so it might be obvious enough that I can change it and
see what the Windows CI thinks.

There isn't yet a packet (or a qSupported feature query) for the
gdb remote serial protocol stub to communicate its watchpoint
capabilities to lldb.  I'll be doing that in a patch right after
this is landed, having debugserver advertise its capability of
AArch64 MASK watchpoints, and have ProcessGDBRemote add
eWatchpointHardwareArmMASK to WatchpointAlgorithms so we can watch
larger than 32-byte requests on Darwin.

I haven't yet tackled WatchpointResource *sharing* by multiple
Watchpoints.  This is all part of the goal, especially when we may
be watching a larger memory range than the user requested, if they
then add another watchpoint next to their first request, it may be
covered by the same WatchpointResource (hardware watchpoint register).
Also one "read" watchpoint and one "write" watchpoint on the same
memory granule need to be handled, making the WatchpointResource
cover all requests.

As WatchpointResources aren't shared among multiple Watchpoints
yet, there's no handling of running the conditions/commands/etc on
multiple Watchpoints when their shared WatchpointResource is hit.
The goal beyond "large watchpoint" is to unify (much more) the
Watchpoint and Breakpoint behavior and commands.  I have a feeling
I may be slowly chipping away at this for a while.

rdar://108234227

jasonmolenda committed:
Finish comment on the trickiness of using MASK watchpoints and
trying to minimize the number of hardware watchpoint registers used
for a single watchpoint.  The pathological cases can lead to watching
giant regions of memory for a small object.

bulbazord commented on lldb/include/lldb/Breakpoint/WatchpointAlgorithms.h:
```diff
@@ -0,0 +1,38 @@
+class WatchpointAlgorithms {
+
+public:
+  static std::vector<lldb::WatchpointResourceSP> AtomizeWatchpointRequest(
```
Can you add some documentation for these static functions? In the form of doxygen comments preferrably

bulbazord commented on lldb/source/Breakpoint/WatchpointAlgorithms.cpp:
```diff
@@ -0,0 +1,146 @@
+    entries =
+        PowerOf2Watchpoints(addr, size,
+                            /* min_byte_size */ 1,
+                            /* max_byte_size */ 2147483648,
```
Is this just INT32_MAX? Probably better to use a constant for clarity.

bulbazord reviewed (commented):
I haven't looked at the algorithm in great detail yet but a lot of the surrounding stuff looks pretty reasonable I think.

jasonmolenda committed:
Address Alex's feedback.

Add doxygen comments for WatchpointAlgorithms.h.
Use INT32_MAX instead of the hardcoded value in
WatchpointAlgorithms.cpp.

jasonmolenda committed:
Fix comment length.

jasonmolenda committed:
Add a test case to TestUnalignedLargeWatchpoint

Watch a 16-byte object and make sure we get a watchpoint hit for
every write to it.  This isn't an unaligned object but it's an easy
place to add this test for now, to make sure we pick up the object
size properly from the variable's type.

JDevlieghere reviewed (approved):
Great job on documenting the code, it's a pleasure to read through. The algorithm makes sense to me and it has solid test coverage. I've left some nits but overall this LGTM.

jasonmolenda commented on lldb/source/Breakpoint/WatchpointAlgorithms.cpp:
```diff
@@ -0,0 +1,146 @@
+      WatchpointHardwareFeature::eWatchpointHardwareArmMASK) {
+    entries =
+        PowerOf2Watchpoints(addr, size,
+                            /* min_byte_size */ 1,
+                            /* max_byte_size */ INT32_MAX,
+                            /* address_byte_size */ arch.GetAddressByteSize());
```
yuck!  :) quick test grep shows 98 for the beautiful spaceful version, 136 for the cruelly spaceless.

jasonmolenda committed:
Address Jonas' feedback.

JDevlieghere commented on lldb/include/lldb/Breakpoint/WatchpointAlgorithms.h:
```diff
@@ -0,0 +1,109 @@
+                                                 uint32_t address_byte_size);
+};
+
+// For the unittests to have access to the individual algorithms
+class WatchpointAlgorithmsTest : public WatchpointAlgorithms {
+public:
+  using WatchpointAlgorithms::PowerOf2Watchpoints;
+};
```
I would've put this in the unittest.

JDevlieghere commented on lldb/source/Breakpoint/WatchpointAlgorithms.cpp:
```diff
@@ -0,0 +1,142 @@
+      WatchpointHardwareFeature::eWatchpointHardwareArmMASK) {
+    entries =
+        PowerOf2Watchpoints(addr, size,
+                            /*min_byte_size*/ 1,
```
Still missing the `=` at the end of the comment.

```/*min_byte_size=*/ 1,```

jasonmolenda commented:
Merged this PR and found two unexpected (and one expected that I forgot to skip a test on) CI bot failures.  I've collected enough logs to debug them locally I think, I reverted this patch from main until I have fixes and I will reland.

DavidSpickett commented:
I didn't have time to review before merge, just looking at this now.

> see what the Windows CI thinks.

I assume you mean the buildbots (lldb isn't built in the pre-commit CI), but I think given:
https://github.com/llvm/llvm-zorg/blob/590f0a62919ad313758362d18a31e7d40255e6e6/buildbot/osuosl/master/config/builders.py#L1342
The tests wouldn't run there anyway.

I don't know specifically why we skip those tests on our bot. Could be that the Windows APIs are target specific like ptrace is, and we just didn't implement it for Windows on Arm yet, could be that the hardware we have doesn't support it.

I'll found out what the status is there.

DavidSpickett commented on lldb/source/Breakpoint/WatchpointAlgorithms.cpp:
```diff
@@ -0,0 +1,142 @@
+                            /*min_byte_size*/ 1,
+                            /*max_byte_size*/ arch.GetAddressByteSize(),
+                            /*address_byte_size*/ arch.GetAddressByteSize());
+  }
```
Seems like it would be simpler to write this as:
```
max_byte_size = 1;
if (supported_features &
    WatchpointHardwareFeature::eWatchpointHardwareArmMASK)
  max_byte_size = INT32_MAX;

std::vector<Region> entries = entries =
        PowerOf2Watchpoints(addr, size,
                            /*min_byte_size=*/ 1,
                            max_byte_size,
                            /*address_byte_size=*/ arch.GetAddressByteSize());
```

DavidSpickett commented on lldb/source/Breakpoint/WatchpointResource.cpp:
```diff
@@ -113,7 +114,8 @@ bool WatchpointResource::ShouldStop(StoppointCallbackContext *context) {
 void WatchpointResource::Dump(Stream *s) const {
-  return; // LWP_TODO
+  s->Printf("addr = 0x%8.8" PRIx64 " size = %zu", m_addr, m_size);
+  return;
```
Don't need a return, unless this thing is supposed to return something?

DavidSpickett commented on lldb/test/API/functionalities/watchpoint/unaligned-large-watchpoint/TestUnalignedLargeWatchpoint.py:
```diff
@@ -0,0 +1,101 @@
+        """Test watching an unaligned region of memory that requires multiple watchpoints."""
+        self.build()
+        self.main_source_file = lldb.SBFileSpec("main.c")
+        (target, process, thread, bkpt) = lldbutil.run_to_source_breakpoint(
```
You don't need ():
target, process, thread, bkpt = ...
Works fine.

DavidSpickett commented on lldb/test/API/functionalities/watchpoint/unaligned-large-watchpoint/TestUnalignedLargeWatchpoint.py:
```diff
@@ -0,0 +1,101 @@
+        if self.TraceOn():
+            self.runCmd("watch list -v")
+
+        c_count = 0
+        reason = self.continue_and_report_stop_reason(process, "continue #%d" % c_count)
+        while reason == lldb.eStopReasonWatchpoint:
+            c_count = c_count + 1
+            reason = self.continue_and_report_stop_reason(
+                process, "continue #%d" % c_count
+            )
+            self.assertLessEqual(c_count, 22)
+
+        self.assertEqual(c_count, 22)
+        self.expect("watchpoint list -v", substrs=["hit_count = 22"])
+        self.assertEqual(wp.GetHitCount(), 22)
```
This could go in some `expected_watchpoint_hits(self, n)` function and be called below as well.

DavidSpickett commented on lldb/test/API/functionalities/watchpoint/unaligned-large-watchpoint/TestUnalignedLargeWatchpoint.py:
```diff
@@ -0,0 +1,101 @@
+
+        # Now try watching a 16 byte variable
+        # (not unaligned, but a good check to do anyway)
+        #
```
stray #

DavidSpickett commented on lldb/test/API/functionalities/watchpoint/unaligned-large-watchpoint/TestUnalignedLargeWatchpoint.py:
```diff
@@ -0,0 +1,101 @@
+        #
+        frame = thread.GetFrameAtIndex(0)
+        err = lldb.SBError()
+        wp = frame.locals["variable"][0].Watch(True, False, True, err)
```
What are the bools here doing?

DavidSpickett reviewed (commented):
Surprised this isn't horribly broken on Linux :)

Just a bunch of nits.

jasonmolenda commented:
Thanks for the great feedback @DavidSpickett sorry I merged it before you had a chance to look.  These are good points, I'll land a cleanup patch in a bit to address them.

jasonmolenda commented:
> I didn't have time to review before merge, just looking at this now.
> 
> > see what the Windows CI thinks.
> 
> I assume you mean the buildbots (lldb isn't built in the pre-commit CI), but I think given: https://github.com/llvm/llvm-zorg/blob/590f0a62919ad313758362d18a31e7d40255e6e6/buildbot/osuosl/master/config/builders.py#L1342 The tests wouldn't run there anyway due to `--skip-category=watchpoint`.

Ah, thanks.  I saw the buildbot fail and assumed it was from the TestUnalignedLargeWatchpoint test I added.  But it was more likely by the fact (Jonas later pointed out) that the compiler on Windows doesn't have `__builtin_clzll` which I was using (I switched to the llvm equiv in the final version).

jasonmolenda commented on lldb/include/lldb/Breakpoint/WatchpointAlgorithms.h:
```diff
@@ -0,0 +1,109 @@
+class WatchpointAlgorithms {
+
+public:
+  /// Convert a user's watchpoint request into an array of memory
```
Thanks, good clarification.

jasonmolenda commented on lldb/include/lldb/Breakpoint/WatchpointAlgorithms.h:
```diff
@@ -0,0 +1,109 @@
+  struct Region {
+    lldb::addr_t addr;
+    size_t size;
+  };
```
I was originally using a std::pair, it's a simple POD tuple used only within the WatchpointAlgorithms class (and the unittest).  If I was passing it outside of these methods I'd agree shouldn't introduce Yet Another Range object.

DavidSpickett commented:
> I don't know specifically why we skip those tests on our bot. Could be that the Windows APIs are target specific like ptrace is, > and we just didn't implement it for Windows on Arm yet, could be that the hardware we have doesn't support it.
> 
> I'll found out what the status is there.

@jasonmolenda, https://github.com/llvm/llvm-project/issues/80665.

They could work but there is work to be done to get there. You don't need to consider it The tests should work on x86 Windows, but there's no bot for that anymore. Maybe someone pops up in future with a report if it's being tested elsewhere.

DavidSpickett commented:
Unlikely even that given the range breakpoint stuff is all on AArch64 anyway.
